File: jena/src/main/scala/eu/ostrzyciel/jelly/convert/jena/riot/JellyFormat.scala
package eu.ostrzyciel.jelly.convert.jena.riot

import eu.ostrzyciel.jelly.convert.jena.riot.JellyLanguage.JELLY
import eu.ostrzyciel.jelly.core.JellyOptions
import eu.ostrzyciel.jelly.core.proto.v1.RdfStreamOptions
import org.apache.jena.riot.{RDFFormat, RDFFormatVariant}

/**
 * Subclass of [[RDFFormatVariant]] to pass Jelly's options to the encoder.
 * @param opt Jelly options
 * @param frameSize size of each RdfStreamFrame, in rows
 * @param enableNamespaceDeclarations whether to include namespace declarations in the output
 * @param delimited whether to write the output as delimited frames. Note: files saved to disk are
 *                  recommended to be delimited, for better interoperability with other implementations.
 *                  In a non-delimited file you can have ONLY ONE FRAME. If the input data is large,
 *                  this will lead to an out-of-memory error. So, this makes sense only for small data.
 *                  **Disable this only if you know what you are doing.**
 */
case class JellyFormatVariant(
  opt: RdfStreamOptions = JellyOptions.smallAllFeatures,
  frameSize: Int = 256,
  enableNamespaceDeclarations: Boolean = false,
  delimited: Boolean = true,
) extends RDFFormatVariant(opt.toString)

/**
 * Pre-defined serialization format variants for Jelly.
 */
object JellyFormat:
  val JELLY_SMALL_STRICT = new RDFFormat(JELLY, JellyFormatVariant(JellyOptions.smallStrict))
  val JELLY_SMALL_GENERALIZED = new RDFFormat(JELLY, JellyFormatVariant(JellyOptions.smallGeneralized))
  val JELLY_SMALL_RDF_STAR = new RDFFormat(JELLY, JellyFormatVariant(JellyOptions.smallRdfStar))
  val JELLY_SMALL_ALL_FEATURES = new RDFFormat(JELLY, JellyFormatVariant(JellyOptions.smallAllFeatures))
  val JELLY_BIG_STRICT = new RDFFormat(JELLY, JellyFormatVariant(JellyOptions.bigStrict))
  val JELLY_BIG_GENERALIZED = new RDFFormat(JELLY, JellyFormatVariant(JellyOptions.bigGeneralized))
  val JELLY_BIG_RDF_STAR = new RDFFormat(JELLY, JellyFormatVariant(JellyOptions.bigRdfStar))
  val JELLY_BIG_ALL_FEATURES = new RDFFormat(JELLY, JellyFormatVariant(JellyOptions.bigAllFeatures))

File: jena-java/src/main/java/eu/neverblink/jelly/convert/jena/riot/JellyFormat.java
package eu.neverblink.jelly.convert.jena.riot;

import eu.neverblink.jelly.core.JellyOptions;
import org.apache.jena.riot.RDFFormat;

/**
 * Pre-defined serialization format variants for Jelly.
 */
public final class JellyFormat {

    private JellyFormat() {}

    public static final RDFFormat JELLY_SMALL_STRICT = new RDFFormat(
        JellyLanguage.JELLY,
        JellyFormatVariant.builder().options(JellyOptions.SMALL_STRICT).build()
    );

    public static final RDFFormat JELLY_SMALL_GENERALIZED = new RDFFormat(
        JellyLanguage.JELLY,
        JellyFormatVariant.builder().options(JellyOptions.SMALL_GENERALIZED).build()
    );

    public static final RDFFormat JELLY_SMALL_RDF_STAR = new RDFFormat(
        JellyLanguage.JELLY,
        JellyFormatVariant.builder().options(JellyOptions.SMALL_RDF_STAR).build()
    );

    public static final RDFFormat JELLY_SMALL_ALL_FEATURES = new RDFFormat(
        JellyLanguage.JELLY,
        JellyFormatVariant.builder().options(JellyOptions.SMALL_ALL_FEATURES).build()
    );

    public static final RDFFormat JELLY_BIG_STRICT = new RDFFormat(
        JellyLanguage.JELLY,
        JellyFormatVariant.builder().options(JellyOptions.BIG_STRICT).build()
    );

    public static final RDFFormat JELLY_BIG_GENERALIZED = new RDFFormat(
        JellyLanguage.JELLY,
        JellyFormatVariant.builder().options(JellyOptions.BIG_GENERALIZED).build()
    );

    public static final RDFFormat JELLY_BIG_RDF_STAR = new RDFFormat(
        JellyLanguage.JELLY,
        JellyFormatVariant.builder().options(JellyOptions.BIG_RDF_STAR).build()
    );

    public static final RDFFormat JELLY_BIG_ALL_FEATURES = new RDFFormat(
        JellyLanguage.JELLY,
        JellyFormatVariant.builder().options(JellyOptions.BIG_ALL_FEATURES).build()
    );
}
